exasol_transformers_extension/udfs/models/base_model_udf.py

```diff
@@ -0,0 +1,196 @@
+        self.base_model = base_model
+        self.tokenizer = tokenizer
+        self.task_name = task_name
+        self.device = None
+        self.cache_dir = None
+        self.last_loaded_model_key = None
+        self.last_loaded_model = None
+        self.last_loaded_tokenizer = None
+        self.last_created_pipeline = None
+
+    def run(self, ctx):
+        device_id = ctx.get_dataframe(1).iloc[0]['device_id']
+        self.device = device_management.get_torch_device(device_id)
+        ctx.reset()
+
+        while True:
+            batch_df = ctx.get_dataframe(num_rows=self.batch_size, start_col=1)
+            if batch_df is None:
+                break
+
+            predictions_df = self.get_predictions_from_batch(batch_df)
+            ctx.emit(predictions_df)
+
+        self.clear_device_memory()
+
+    def get_predictions_from_batch(self, batch_df: pd.DataFrame) -> pd.DataFrame:
+        """
+        Perform separate predictions for each model in the dataframe.
+
+        :param batch_df: A batch of dataframe retrieved from context
+
+        :return: Prediction results of the corresponding batched dataframe
+        """
+        result_df_list = []
+        for model_df in \
+                self.extract_unique_model_dataframes_from_batch(batch_df):
+            for param_based_model_df in \
+                    self.extract_unique_param_based_dataframes(model_df):
+                result_df = self.get_prediction(param_based_model_df)
+                result_df_list.append(result_df)
+
+        result_df = pd.concat(result_df_list)
+        return result_df
+
+    def extract_unique_model_dataframes_from_batch(
+            self, batch_df: pd.DataFrame) -> Iterator[pd.DataFrame]:
+        """
+        Extract unique model dataframes with the same model_name, bucketfs_conn,
+        and sub_dir from the dataframe. If the extracted model is not cached,
+        it is loaded into the cache before performing the prediction.
+
+        :param batch_df: A batch of dataframe retrieved from context
+
+        :return: Unique model dataframes having same model_name,
+        bucketfs_connection, and sub_dir
+        """
+
+        unique_values = dataframe_operations.get_unique_values(
+            batch_df, constants.ORDERED_COLUMNS, sort=True)
+        for model_name, bucketfs_conn, sub_dir in unique_values:
+            model_df = batch_df[
+                (batch_df['model_name'] == model_name) &
+                (batch_df['bucketfs_conn'] == bucketfs_conn) &
+                (batch_df['sub_dir'] == sub_dir)]
+
+            current_model_key = (bucketfs_conn, sub_dir, model_name)
+            if self.last_loaded_model_key != current_model_key:
+                self.set_cache_dir(model_df)
+                self.clear_device_memory()
+                self.load_models(model_name)
+                self.last_loaded_model_key = current_model_key
+
+            yield model_df
+
+    def extract_unique_param_based_dataframes(
```

Comment:
Should be this method abstract, because tbe default implementation is not complex and with it , new implementation might not know about it